pkg/controllers/management/eks/eks_cluster_handler.go

```diff
@@ -0,0 +1,561 @@
+		e.clusterEngqueueAfter(cluster.Name, 20*time.Second)
+		return cluster, nil
+	case "active":
+		if aws.BoolValue(cluster.Spec.EKSConfig.Imported) {
+			if cluster.Annotations == nil || cluster.Annotations[importedAnno] != "true" {
+				// imported annotation is not set- need to get spec from upstream cluster
+				var eksClusterConfigStructured eksv1.EKSClusterConfig
+				if err := runtime.DefaultUnstructuredConverter.FromUnstructured(eksClusterConfigDynamic.Object, &eksClusterConfigStructured); err != nil {
+					return cluster, fmt.Errorf("failed to import cluster [%s]", cluster.Name)
+				}
+
+				// copy EKS cluster config spec to cluster and set imported annotation
+				cluster = cluster.DeepCopy()
+				cluster.Spec.EKSConfig = &eksClusterConfigStructured.Spec
+				cluster.Status.AppliedSpec = cluster.Spec
+				if cluster.Annotations == nil {
+					cluster.Annotations = make(map[string]string)
+				}
+				cluster.Annotations[importedAnno] = "true"
+				cluster, err := e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+
+		if aws.BoolValue(cluster.Spec.EKSConfig.Imported) {
+			// cluster spec has been copied and cluster is no longer pending
+			if !apimgmtv3.ClusterConditionPending.IsTrue(cluster) {
+				cluster = cluster.DeepCopy()
+				apimgmtv3.ClusterConditionPending.True(cluster)
+				cluster, err = e.clusterClient.Update(cluster)
+			}
+		}
+
+		if !apimgmtv3.ClusterConditionProvisioned.IsTrue(cluster) {
+			// all provisioning has fnished
+			cluster = cluster.DeepCopy()
+			apimgmtv3.ClusterConditionProvisioned.True(cluster)
+			cluster, err = e.clusterClient.Update(cluster)
+			if err != nil {
+				return cluster, err
+			}
+		}
+
+		if cluster.Status.ServiceAccountToken == "" || cluster.Status.APIEndpoint == "" {
+			// service account token and API endpoint are need to deploy cluster agent, should be able to retrieve
+			// from secret
+			caSecret, err := e.secretsCache.Get(globalNS, cluster.Name)
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					return cluster, err
+				}
+				logrus.Infof("waiting for cluster [%s] data needed to generate service account token", cluster.Name)
+				e.clusterEngqueueAfter(cluster.Name, 20*time.Second)
+				return cluster, nil
+			}
+
+			logrus.Infof("generating service account token for cluster [%s]", cluster.Name)
+			sess, err := e.startAWSSession(cluster.Spec.EKSConfig.AmazonCredentialSecret)
+			if err != nil {
+				return cluster, nil
+			}
+
+			endpoint := string(caSecret.Data["endpoint"])
+			ca := string(caSecret.Data["ca"])
+			saToken, err := generateSAToken(sess, cluster.Spec.EKSConfig.DisplayName, endpoint, ca)
+			if err != nil {
+				return cluster, err
+			}
+
+			cluster = cluster.DeepCopy()
+			cluster.Status.APIEndpoint = endpoint
+			cluster.Status.CACert = ca
+			cluster.Status.ServiceAccountToken = saToken
+			return e.clusterClient.Update(cluster)
+		}
+
+		// check for unauthorized error
```

Comment:
Why is this check so far below ? Because we just got the SA token/endpoint ?

Reply:
The logic above covers clusters that have just become active for the first time. This block could be relevant anytime as the eks-operator attempts to create a session prior to any update.